Unit tests must not require credentials in environment

Creating an actual `Client` / `AsyncClient` instance without passing in credentials raises an error.

Better practice for anything not directly testing the client is to use `mock.create_autospec` to create
a dummy client.

```bash
$ git remote -v
origin	git@github.com:googleapis/python-firestore (fetch)
origin	git@github.com:googleapis/python-firestore (push)
$ git log -1
commit 51c92aa45156c5462f35991f65c02dcac49aacc3 (HEAD -> main, tag: v2.3.2, origin/main)
Author: release-please[bot] <55107282+release-please[bot]@users.noreply.github.com>
Date:   Tue Sep 14 14:53:30 2021 -0700

    chore: release 2.3.2 (#441)
    
    Co-authored-by: release-please[bot] <55107282+release-please[bot]@users.noreply.github.com>
$ git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean
$ env | grep GOOGLE || echo "No Credentials"
No Credentials
$ nox -e unit-3.6
...
________________ TestScheduling.test_operation_retry_scheduling ________________

self = <tests.unit.v1.test_bulk_writer.TestScheduling testMethod=test_operation_retry_scheduling>

    def test_operation_retry_scheduling(self):
        now = datetime.datetime.now()
        one_second_from_now = now + datetime.timedelta(seconds=1)
    
>       db = Client()

tests/unit/v1/test_bulk_writer.py:574: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
google/cloud/firestore_v1/client.py:99: in __init__
    client_options=client_options,
google/cloud/firestore_v1/base_client.py:129: in __init__
    _http=None,
.nox/unit-3-6/lib/python3.6/site-packages/google/cloud/client.py:249: in __init__
    _ClientProjectMixin.__init__(self, project=project)
.nox/unit-3-6/lib/python3.6/site-packages/google/cloud/client.py:201: in __init__
    project = self._determine_default(project)
.nox/unit-3-6/lib/python3.6/site-packages/google/cloud/client.py:216: in _determine_default
    return _determine_default_project(project)
.nox/unit-3-6/lib/python3.6/site-packages/google/cloud/_helpers.py:186: in _determine_default_project
    _, project = google.auth.default()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

scopes = None, request = None, quota_project_id = None

    def default(scopes=None, request=None, quota_project_id=None):
        """Gets the default credentials for the current environment.
    
        `Application Default Credentials`_ provides an easy way to obtain
        credentials to call Google APIs for server-to-server or local applications.
        This function acquires credentials from the environment in the following
        order:
    
        1. If the environment variable ``GOOGLE_APPLICATION_CREDENTIALS`` is set
           to the path of a valid service account JSON private key file, then it is
           loaded and returned. The project ID returned is the project ID defined
           in the service account file if available (some older files do not
           contain project ID information).
        2. If the `Google Cloud SDK`_ is installed and has application default
           credentials set they are loaded and returned.
    
           To enable application default credentials with the Cloud SDK run::
    
                gcloud auth application-default login
    
           If the Cloud SDK has an active project, the project ID is returned. The
           active project can be set using::
    
                gcloud config set project
    
        3. If the application is running in the `App Engine standard environment`_
           (first generation) then the credentials and project ID from the
           `App Identity Service`_ are used.
        4. If the application is running in `Compute Engine`_ or `Cloud Run`_ or
           the `App Engine flexible environment`_ or the `App Engine standard
           environment`_ (second generation) then the credentials and project ID
           are obtained from the `Metadata Service`_.
        5. If no credentials are found,
           :class:`~google.auth.exceptions.DefaultCredentialsError` will be raised.
    
        .. _Application Default Credentials: https://developers.google.com\
                /identity/protocols/application-default-credentials
        .. _Google Cloud SDK: https://cloud.google.com/sdk
        .. _App Engine standard environment: https://cloud.google.com/appengine
        .. _App Identity Service: https://cloud.google.com/appengine/docs/python\
                /appidentity/
        .. _Compute Engine: https://cloud.google.com/compute
        .. _App Engine flexible environment: https://cloud.google.com\
                /appengine/flexible
        .. _Metadata Service: https://cloud.google.com/compute/docs\
                /storing-retrieving-metadata
        .. _Cloud Run: https://cloud.google.com/run
    
        Example::
    
            import google.auth
    
            credentials, project_id = google.auth.default()
    
        Args:
            scopes (Sequence[str]): The list of scopes for the credentials. If
                specified, the credentials will automatically be scoped if
                necessary.
            request (google.auth.transport.Request): An object used to make
                HTTP requests. This is used to detect whether the application
                is running on Compute Engine. If not specified, then it will
                use the standard library http client to make requests.
            quota_project_id (Optional[str]):  The project ID used for
                quota and billing.
        Returns:
            Tuple[~google.auth.credentials.Credentials, Optional[str]]:
                the current environment's credentials and project ID. Project ID
                may be None, which indicates that the Project ID could not be
                ascertained from the environment.
    
        Raises:
            ~google.auth.exceptions.DefaultCredentialsError:
                If no credentials were found, or if the credentials found were
                invalid.
        """
        from google.auth.credentials import with_scopes_if_required
    
        explicit_project_id = os.environ.get(
            environment_vars.PROJECT, os.environ.get(environment_vars.LEGACY_PROJECT)
        )
    
        checkers = (
            _get_explicit_environ_credentials,
            _get_gcloud_sdk_credentials,
            _get_gae_credentials,
            lambda: _get_gce_credentials(request),
        )
    
        for checker in checkers:
            credentials, project_id = checker()
            if credentials is not None:
                credentials = with_scopes_if_required(credentials, scopes)
                if quota_project_id:
                    credentials = credentials.with_quota_project(quota_project_id)
    
                effective_project_id = explicit_project_id or project_id
                if not effective_project_id:
                    _LOGGER.warning(
                        "No project ID could be determined. Consider running "
                        "`gcloud config set project` or setting the %s "
                        "environment variable",
                        environment_vars.PROJECT,
                    )
                return credentials, effective_project_id
    
>       raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
E       google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started

.nox/unit-3-6/lib/python3.6/site-packages/google/auth/_default.py:356: DefaultCredentialsError
=========================== short test summary info ============================
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_cannot_add_after_close
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_create_calls_send_correctly
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_create_retriable
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_default_error_handler
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_delete_calls_send_correctly
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_delete_retriable
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_delete_separates_batch
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_ensure_sending_repeatedly_callable
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_flush_close_repeatedly_callable
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_flush_sends_all_queued_batches
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_flush_sends_in_progress
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_handles_errors_and_successes_correctly
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_invokes_error_callbacks_successfully
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_invokes_error_callbacks_successfully_multiple_retries
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_invokes_success_callbacks_successfully
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_multiple_flushes
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_separates_same_document
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_separates_same_document_different_operation
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_serial_calls_send_correctly
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_set_calls_send_correctly
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_set_retriable
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_update_calls_send_correctly
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_update_raises_with_bad_option
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_update_retriable
FAILED tests/unit/v1/test_bulk_writer.py::TestSyncBulkWriter::test_update_separates_batch
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_cannot_add_after_close
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_create_calls_send_correctly
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_create_retriable
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_default_error_handler
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_delete_calls_send_correctly
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_delete_retriable
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_delete_separates_batch
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_ensure_sending_repeatedly_callable
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_flush_close_repeatedly_callable
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_flush_sends_all_queued_batches
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_flush_sends_in_progress
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_handles_errors_and_successes_correctly
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_invokes_error_callbacks_successfully
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_invokes_error_callbacks_successfully_multiple_retries
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_invokes_success_callbacks_successfully
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_multiple_flushes
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_separates_same_document
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_separates_same_document_different_operation
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_serial_calls_send_correctly
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_set_calls_send_correctly
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_set_retriable
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_update_calls_send_correctly
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_update_raises_with_bad_option
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_update_retriable
FAILED tests/unit/v1/test_bulk_writer.py::TestAsyncBulkWriter::test_update_separates_batch
FAILED tests/unit/v1/test_bulk_writer.py::TestScheduling::test_max_in_flight_honored
FAILED tests/unit/v1/test_bulk_writer.py::TestScheduling::test_operation_retry_scheduling
============================= 52 failed in 10.50s ==============================

--- a/testing/constraints-3.6.txt
+++ b/testing/constraints-3.6.txt
@@ -8,4 +8,5 @@
 google-api-core==1.26.0
 google-cloud-core==1.4.1
 proto-plus==1.10.0
+protobuf==3.12.0  # transitive from `google-api-core`
 google-auth==1.24.0  # TODO: remove when google-auth>=1.25.0 is required through google-api-core

--- a/tests/unit/v1/test_bulk_writer.py
+++ b/tests/unit/v1/test_bulk_writer.py
@@ -82,10 +82,12 @@ class NoSendBulkWriter(BulkWriter):
     def _instantiate_executor(self):
         return FakeThreadPoolExecutor()
 
+
 def _make_credentials():
     from google.auth.credentials import Credentials
 
     return mock.create_autospec(Credentials, project_id="project-id")
+
 
 class _SyncClientMixin:
     """Mixin which helps a `_BaseBulkWriterTests` subclass simulate usage of
@@ -106,10 +108,11 @@ class _AsyncClientMixin:
 
 
 class _BaseBulkWriterTests:
-
     @staticmethod
     def _get_document_reference(
-        client: BaseClient, collection_name: Optional[str] = "col", id: Optional[str] = None,
+        client: BaseClient,
+        collection_name: Optional[str] = "col",
+        id: Optional[str] = None,
     ) -> Type:
         return client.collection(collection_name).document(id)
 
@@ -481,9 +484,7 @@ class _BaseBulkWriterTests:
 
     def test_serial_calls_send_correctly(self):
         client = self._make_client()
-        bw = NoSendBulkWriter(
-            client, options=BulkWriterOptions(mode=SendMode.serial)
-        )
+        bw = NoSendBulkWriter(client, options=BulkWriterOptions(mode=SendMode.serial))
         for ref, data in self._doc_iter(client, 101):
             bw.create(ref, data)
         bw.flush()
@@ -579,7 +580,6 @@ class TestAsyncBulkWriter(
 
 
 class TestScheduling(unittest.TestCase):
-
     @staticmethod
     def _make_client() -> Client:
         return Client(credentials=_make_credentials(), project="project-id")

--- a/tests/unit/v1/test_bulk_writer.py
+++ b/tests/unit/v1/test_bulk_writer.py
@@ -18,6 +18,7 @@ from typing import List, NoReturn, Optional, Tuple, Type
 
 from google.rpc import status_pb2
 import aiounittest  # type: ignore
+import mock
 
 from google.cloud.firestore_v1._helpers import build_timestamp, ExistsOption
 from google.cloud.firestore_v1.async_client import AsyncClient
@@ -81,36 +82,41 @@ class NoSendBulkWriter(BulkWriter):
     def _instantiate_executor(self):
         return FakeThreadPoolExecutor()
 
+def _make_credentials():
+    from google.auth.credentials import Credentials
+
+    return mock.create_autospec(Credentials, project_id="project-id")
 
 class _SyncClientMixin:
     """Mixin which helps a `_BaseBulkWriterTests` subclass simulate usage of
     synchronous Clients, Collections, DocumentReferences, etc."""
 
-    def _get_client_class(self) -> Type:
-        return Client
+    @staticmethod
+    def _make_client() -> Client:
+        return Client(credentials=_make_credentials(), project="project-id")
 
 
 class _AsyncClientMixin:
     """Mixin which helps a `_BaseBulkWriterTests` subclass simulate usage of
     AsyncClients, AsyncCollections, AsyncDocumentReferences, etc."""
 
-    def _get_client_class(self) -> Type:
-        return AsyncClient
+    @staticmethod
+    def _make_client() -> AsyncClient:
+        return AsyncClient(credentials=_make_credentials(), project="project-id")
 
 
 class _BaseBulkWriterTests:
-    def setUp(self):
-        self.client: BaseClient = self._get_client_class()()
 
+    @staticmethod
     def _get_document_reference(
-        self, collection_name: Optional[str] = "col", id: Optional[str] = None,
+        client: BaseClient, collection_name: Optional[str] = "col", id: Optional[str] = None,
     ) -> Type:
-        return self.client.collection(collection_name).document(id)
+        return client.collection(collection_name).document(id)
 
-    def _doc_iter(self, num: int, ids: Optional[List[str]] = None):
+    def _doc_iter(self, client, num: int, ids: Optional[List[str]] = None):
         for _ in range(num):
             id: Optional[str] = ids[_] if ids else None
-            yield self._get_document_reference(id=id), {"id": _}
+            yield self._get_document_reference(client, id=id), {"id": _}
 
     def _verify_bw_activity(self, bw: BulkWriter, counts: List[Tuple[int, int]]):
         """
@@ -143,8 +149,9 @@ class _BaseBulkWriterTests:
         self.assertEqual(len(bw._operations), 0)
 
     def test_create_calls_send_correctly(self):
-        bw = NoSendBulkWriter(self.client)
-        for ref, data in self._doc_iter(101):
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
+        for ref, data in self._doc_iter(client, 101):
             bw.create(ref, data)
         bw.flush()
         # Full batches with 20 items should have been sent 5 times, and a 1-item
@@ -152,8 +159,9 @@ class _BaseBulkWriterTests:
         self._verify_bw_activity(bw, [(20, 5,), (1, 1,)])
 
     def test_delete_calls_send_correctly(self):
-        bw = NoSendBulkWriter(self.client)
-        for ref, _ in self._doc_iter(101):
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
+        for ref, _ in self._doc_iter(client, 101):
             bw.delete(ref)
         bw.flush()
         # Full batches with 20 items should have been sent 5 times, and a 1-item
@@ -161,8 +169,9 @@ class _BaseBulkWriterTests:
         self._verify_bw_activity(bw, [(20, 5,), (1, 1,)])
 
     def test_delete_separates_batch(self):
-        bw = NoSendBulkWriter(self.client)
-        ref = self._get_document_reference(id="asdf")
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
+        ref = self._get_document_reference(client, id="asdf")
         bw.create(ref, {})
         bw.delete(ref)
         bw.flush()
@@ -170,8 +179,9 @@ class _BaseBulkWriterTests:
         self._verify_bw_activity(bw, [(1, 2,)])
 
     def test_set_calls_send_correctly(self):
-        bw = NoSendBulkWriter(self.client)
-        for ref, data in self._doc_iter(101):
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
+        for ref, data in self._doc_iter(client, 101):
             bw.set(ref, data)
         bw.flush()
         # Full batches with 20 items should have been sent 5 times, and a 1-item
@@ -179,8 +189,9 @@ class _BaseBulkWriterTests:
         self._verify_bw_activity(bw, [(20, 5,), (1, 1,)])
 
     def test_update_calls_send_correctly(self):
-        bw = NoSendBulkWriter(self.client)
-        for ref, data in self._doc_iter(101):
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
+        for ref, data in self._doc_iter(client, 101):
             bw.update(ref, data)
         bw.flush()
         # Full batches with 20 items should have been sent 5 times, and a 1-item
@@ -188,8 +199,9 @@ class _BaseBulkWriterTests:
         self._verify_bw_activity(bw, [(20, 5,), (1, 1,)])
 
     def test_update_separates_batch(self):
-        bw = NoSendBulkWriter(self.client)
-        ref = self._get_document_reference(id="asdf")
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
+        ref = self._get_document_reference(client, id="asdf")
         bw.create(ref, {})
         bw.update(ref, {"field": "value"})
         bw.flush()
@@ -198,7 +210,8 @@ class _BaseBulkWriterTests:
         self._verify_bw_activity(bw, [(1, 2,)])
 
     def test_invokes_success_callbacks_successfully(self):
-        bw = NoSendBulkWriter(self.client)
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
         bw._fail_indices = []
         bw._sent_batches = 0
         bw._sent_documents = 0
@@ -218,7 +231,7 @@ class _BaseBulkWriterTests:
         bw.on_write_result(_on_write)
         bw.on_batch_result(_on_batch)
 
-        for ref, data in self._doc_iter(101):
+        for ref, data in self._doc_iter(client, 101):
             bw.create(ref, data)
         bw.flush()
 
@@ -227,7 +240,8 @@ class _BaseBulkWriterTests:
         self.assertEqual(len(bw._operations), 0)
 
     def test_invokes_error_callbacks_successfully(self):
-        bw = NoSendBulkWriter(self.client)
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
         # First document in each batch will "fail"
         bw._fail_indices = [0]
         bw._sent_batches = 0
@@ -253,7 +267,7 @@ class _BaseBulkWriterTests:
         bw.on_write_result(_on_write)
         bw.on_write_error(_on_error)
 
-        for ref, data in self._doc_iter(1):
+        for ref, data in self._doc_iter(client, 1):
             bw.create(ref, data)
         bw.flush()
 
@@ -263,8 +277,9 @@ class _BaseBulkWriterTests:
         self.assertEqual(len(bw._operations), 0)
 
     def test_invokes_error_callbacks_successfully_multiple_retries(self):
+        client = self._make_client()
         bw = NoSendBulkWriter(
-            self.client, options=BulkWriterOptions(retry=BulkRetry.immediate),
+            client, options=BulkWriterOptions(retry=BulkRetry.immediate),
         )
         # First document in each batch will "fail"
         bw._fail_indices = [0]
@@ -291,7 +306,7 @@ class _BaseBulkWriterTests:
         bw.on_write_result(_on_write)
         bw.on_write_error(_on_error)
 
-        for ref, data in self._doc_iter(2):
+        for ref, data in self._doc_iter(client, 2):
             bw.create(ref, data)
         bw.flush()
 
@@ -301,8 +316,9 @@ class _BaseBulkWriterTests:
         self.assertEqual(len(bw._operations), 0)
 
     def test_default_error_handler(self):
+        client = self._make_client()
         bw = NoSendBulkWriter(
-            self.client, options=BulkWriterOptions(retry=BulkRetry.immediate),
+            client, options=BulkWriterOptions(retry=BulkRetry.immediate),
         )
         bw._attempts = 0
 
@@ -314,14 +330,15 @@ class _BaseBulkWriterTests:
 
         # First document in each batch will "fail"
         bw._fail_indices = [0]
-        for ref, data in self._doc_iter(1):
+        for ref, data in self._doc_iter(client, 1):
             bw.create(ref, data)
         bw.flush()
         self.assertEqual(bw._attempts, 15)
 
     def test_handles_errors_and_successes_correctly(self):
+        client = self._make_client()
         bw = NoSendBulkWriter(
-            self.client, options=BulkWriterOptions(retry=BulkRetry.immediate),
+            client, options=BulkWriterOptions(retry=BulkRetry.immediate),
         )
         # First document in each batch will "fail"
         bw._fail_indices = [0]
@@ -348,7 +365,7 @@ class _BaseBulkWriterTests:
         bw.on_write_result(_on_write)
         bw.on_write_error(_on_error)
 
-        for ref, data in self._doc_iter(40):
+        for ref, data in self._doc_iter(client, 40):
             bw.create(ref, data)
         bw.flush()
 
@@ -359,8 +376,9 @@ class _BaseBulkWriterTests:
         self.assertEqual(len(bw._operations), 0)
 
     def test_create_retriable(self):
+        client = self._make_client()
         bw = NoSendBulkWriter(
-            self.client, options=BulkWriterOptions(retry=BulkRetry.immediate),
+            client, options=BulkWriterOptions(retry=BulkRetry.immediate),
         )
         # First document in each batch will "fail"
         bw._fail_indices = [0]
@@ -376,7 +394,7 @@ class _BaseBulkWriterTests:
 
         bw.on_write_error(_on_error)
 
-        for ref, data in self._doc_iter(1):
+        for ref, data in self._doc_iter(client, 1):
             bw.create(ref, data)
         bw.flush()
 
@@ -384,8 +402,9 @@ class _BaseBulkWriterTests:
         self.assertEqual(len(bw._operations), 0)
 
     def test_delete_retriable(self):
+        client = self._make_client()
         bw = NoSendBulkWriter(
-            self.client, options=BulkWriterOptions(retry=BulkRetry.immediate),
+            client, options=BulkWriterOptions(retry=BulkRetry.immediate),
         )
         # First document in each batch will "fail"
         bw._fail_indices = [0]
@@ -401,7 +420,7 @@ class _BaseBulkWriterTests:
 
         bw.on_write_error(_on_error)
 
-        for ref, _ in self._doc_iter(1):
+        for ref, _ in self._doc_iter(client, 1):
             bw.delete(ref)
         bw.flush()
 
@@ -409,8 +428,9 @@ class _BaseBulkWriterTests:
         self.assertEqual(len(bw._operations), 0)
 
     def test_set_retriable(self):
+        client = self._make_client()
         bw = NoSendBulkWriter(
-            self.client, options=BulkWriterOptions(retry=BulkRetry.immediate),
+            client, options=BulkWriterOptions(retry=BulkRetry.immediate),
         )
         # First document in each batch will "fail"
         bw._fail_indices = [0]
@@ -426,7 +446,7 @@ class _BaseBulkWriterTests:
 
         bw.on_write_error(_on_error)
 
-        for ref, data in self._doc_iter(1):
+        for ref, data in self._doc_iter(client, 1):
             bw.set(ref, data)
         bw.flush()
 
@@ -434,8 +454,9 @@ class _BaseBulkWriterTests:
         self.assertEqual(len(bw._operations), 0)
 
     def test_update_retriable(self):
+        client = self._make_client()
         bw = NoSendBulkWriter(
-            self.client, options=BulkWriterOptions(retry=BulkRetry.immediate),
+            client, options=BulkWriterOptions(retry=BulkRetry.immediate),
         )
         # First document in each batch will "fail"
         bw._fail_indices = [0]
@@ -451,7 +472,7 @@ class _BaseBulkWriterTests:
 
         bw.on_write_error(_on_error)
 
-        for ref, data in self._doc_iter(1):
+        for ref, data in self._doc_iter(client, 1):
             bw.update(ref, data)
         bw.flush()
 
@@ -459,10 +480,11 @@ class _BaseBulkWriterTests:
         self.assertEqual(len(bw._operations), 0)
 
     def test_serial_calls_send_correctly(self):
+        client = self._make_client()
         bw = NoSendBulkWriter(
-            self.client, options=BulkWriterOptions(mode=SendMode.serial)
+            client, options=BulkWriterOptions(mode=SendMode.serial)
         )
-        for ref, data in self._doc_iter(101):
+        for ref, data in self._doc_iter(client, 101):
             bw.create(ref, data)
         bw.flush()
         # Full batches with 20 items should have been sent 5 times, and a 1-item
@@ -470,8 +492,9 @@ class _BaseBulkWriterTests:
         self._verify_bw_activity(bw, [(20, 5,), (1, 1,)])
 
     def test_separates_same_document(self):
-        bw = NoSendBulkWriter(self.client)
-        for ref, data in self._doc_iter(2, ["same-id", "same-id"]):
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
+        for ref, data in self._doc_iter(client, 2, ["same-id", "same-id"]):
             bw.create(ref, data)
         bw.flush()
         # Seeing the same document twice should lead to separate batches
@@ -479,8 +502,9 @@ class _BaseBulkWriterTests:
         self._verify_bw_activity(bw, [(1, 2,)])
 
     def test_separates_same_document_different_operation(self):
-        bw = NoSendBulkWriter(self.client)
-        for ref, data in self._doc_iter(1, ["same-id"]):
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
+        for ref, data in self._doc_iter(client, 1, ["same-id"]):
             bw.create(ref, data)
             bw.set(ref, data)
         bw.flush()
@@ -489,47 +513,54 @@ class _BaseBulkWriterTests:
         self._verify_bw_activity(bw, [(1, 2,)])
 
     def test_ensure_sending_repeatedly_callable(self):
-        bw = NoSendBulkWriter(self.client)
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
         bw._is_sending = True
         bw._ensure_sending()
 
     def test_flush_close_repeatedly_callable(self):
-        bw = NoSendBulkWriter(self.client)
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
         bw.flush()
         bw.flush()
         bw.close()
 
     def test_flush_sends_in_progress(self):
-        bw = NoSendBulkWriter(self.client)
-        bw.create(self._get_document_reference(), {"whatever": "you want"})
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
+        bw.create(self._get_document_reference(client), {"whatever": "you want"})
         bw.flush()
         self._verify_bw_activity(bw, [(1, 1,)])
 
     def test_flush_sends_all_queued_batches(self):
-        bw = NoSendBulkWriter(self.client)
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
         for _ in range(2):
-            bw.create(self._get_document_reference(), {"whatever": "you want"})
+            bw.create(self._get_document_reference(client), {"whatever": "you want"})
             bw._queued_batches.append(bw._operations)
             bw._reset_operations()
         bw.flush()
         self._verify_bw_activity(bw, [(1, 2,)])
 
     def test_cannot_add_after_close(self):
-        bw = NoSendBulkWriter(self.client)
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
         bw.close()
         self.assertRaises(Exception, bw._verify_not_closed)
 
     def test_multiple_flushes(self):
-        bw = NoSendBulkWriter(self.client)
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
         bw.flush()
         bw.flush()
 
     def test_update_raises_with_bad_option(self):
-        bw = NoSendBulkWriter(self.client)
+        client = self._make_client()
+        bw = NoSendBulkWriter(client)
         self.assertRaises(
             ValueError,
             bw.update,
-            self._get_document_reference("id"),
+            self._get_document_reference(client, "id"),
             {},
             option=ExistsOption(exists=True),
         )
@@ -548,8 +579,13 @@ class TestAsyncBulkWriter(
 
 
 class TestScheduling(unittest.TestCase):
+
+    @staticmethod
+    def _make_client() -> Client:
+        return Client(credentials=_make_credentials(), project="project-id")
+
     def test_max_in_flight_honored(self):
-        bw = NoSendBulkWriter(Client())
+        bw = NoSendBulkWriter(self._make_client())
         # Calling this method sets up all the internal timekeeping machinery
         bw._rate_limiter.take_tokens(20)
 
@@ -571,7 +607,7 @@ class TestScheduling(unittest.TestCase):
         now = datetime.datetime.now()
         one_second_from_now = now + datetime.timedelta(seconds=1)
 
-        db = Client()
+        db = self._make_client()
         operation = BulkWriterCreateOperation(
             reference=db.collection("asdf").document("asdf"),
             document_data={"does.not": "matter"},

--- a/tests/unit/v1/test_bundle.py
+++ b/tests/unit/v1/test_bundle.py
@@ -419,6 +419,7 @@ class TestBundleBuilder(_CollectionQueryMixin, unittest.TestCase):
         """Some SDKs (Node) serialize Timestamp values to
         '{"seconds": 123, "nanos": 456}', instead of an ISO-formatted string.
         This tests deserialization from that format."""
+        from google.protobuf.json_format import ParseError
 
         client = _test_helpers.make_client(project_name="fir-bundles-test")
 
@@ -441,7 +442,9 @@ class TestBundleBuilder(_CollectionQueryMixin, unittest.TestCase):
         )
 
         self.assertRaises(
-            ValueError, _helpers.deserialize_bundle, _serialized, client=client,
+            (ValueError, ParseError),  # protobuf 3.18.0 raises ParseError
+            _helpers.deserialize_bundle,
+            _serialized, client=client,
         )
 
         # The following assertions would test deserialization of NodeJS bundles

--- a/tests/unit/v1/test_bundle.py
+++ b/tests/unit/v1/test_bundle.py
@@ -444,7 +444,8 @@ class TestBundleBuilder(_CollectionQueryMixin, unittest.TestCase):
         self.assertRaises(
             (ValueError, ParseError),  # protobuf 3.18.0 raises ParseError
             _helpers.deserialize_bundle,
-            _serialized, client=client,
+            _serialized,
+            client=client,
         )
 
         # The following assertions would test deserialization of NodeJS bundles

